protocols/direct-funding.go

```diff
@@ -0,0 +1,198 @@
+		if len(effects) > 0 {
+			// todo: effects as SideEffects{}
+			return s, NoSideEffects, WaitingForCompleteFunding, nil
+		}
+	}
+
+	if !fundingComplete {
+		return s, NoSideEffects, WaitingForCompleteFunding, nil
+	}
+
+	// Postfunding
+	if !s.PostFundSigned[s.MyIndex] {
+		// todo: []string{SignPostFundEffect(s.ChannelId)} as SideEffects{}
+		return s, NoSideEffects, WaitingForCompletePostFund, nil
+	}
+
+	if !s.PostfundComplete() {
+		return s, NoSideEffects, WaitingForCompletePostFund, nil
+	}
+
+	// Completion
+	// todo: []string{"Objective" + s.ChannelId.String() + "complete"} as SideEffects{}
+	return s, NoSideEffects, WaitingForNothing, nil
+}
+
+func (s DirectFundingObjectiveState) Update(event ObjectiveEvent) Objective {
+	return s
+}
+
+func (s DirectFundingObjectiveState) Reject() Objective {
+	updated := s.Clone()
+	updated.Status = Rejected
+	return updated
+}
+
+func (s DirectFundingObjectiveState) Approve() Objective {
+	updated := s.Clone()
+	// todo: consider case of s.Status == Rejected
+	updated.Status = Approved
+
+	return updated
+}
+
+func (s DirectFundingObjectiveState) Id() ObjectiveId {
+	return ObjectiveId("DirectFunding-" + s.ChannelId.String())
+}
+
+func InitializeDirectFundingObjectiveState(initialState state.State) (DirectFundingObjectiveState, error) {
+	var init DirectFundingObjectiveState
+	var err error
+
+	init.Status = Unapproved
+	init.ChannelId, err = initialState.ChannelId()
+	if err != nil {
+		return init, err
+	}
+	for i, v := range initialState.Participants {
+		init.ParticipantIndex[v] = uint(i)
+	}
+
+	init.ExpectedStates[0] = initialState
+
+	fixed := initialState.FixedPart()
+	init.ExpectedStates[1].ChainId = fixed.ChainId
+	init.ExpectedStates[1].Participants = fixed.Participants
+	init.ExpectedStates[1].ChannelNonce = fixed.ChannelNonce
+	init.ExpectedStates[1].AppDefinition = fixed.AppDefinition
+	init.ExpectedStates[1].ChallengeDuration = fixed.ChallengeDuration
+
+	init.ExpectedStates[1].Outcome = initialState.Outcome
+	init.ExpectedStates[1].AppData = initialState.AppData
+	init.ExpectedStates[1].IsFinal = false
+	init.ExpectedStates[1].TurnNum = big.NewInt(1)
+
+	// init.MyIndex						// requires myAddress
+
+	// init.MyDepositSafetyThreshold	// \
+	// init.MyDepositTarget				//  -> further fcn inputs?
+	// init.FullyFundedThreshold		// /
+
+	init.PostFundSigned = make([]bool, len(initialState.Participants))
+	init.OnChainHolding = types.Holdings{}
+
+	return init, nil
+}
+
+func (s DirectFundingObjectiveState) FundingUpdated(amount types.Holdings, finalized bool) (DirectFundingObjectiveState, error) {
+	updated := s.Clone()
+	updated.OnChainHolding.Add(amount)
+	return updated, nil
+}
+
+// SignatureReceived updates the objective's cache of which participants have signed which states
+func (s DirectFundingObjectiveState) SignatureRecieved(signature state.Signature, turnNum int) (DirectFundingObjectiveState, error) {
```

Comment:
I think we want to combine all of these named methods into the `Update` method. (You may want to keep them as private methods which are executed inside of a loop).